File: hacking_with_perl/HTML5_Cross-Origin_Messaging.pl
#!/usr/bin/perl
use strict;
use warnings;
use LWP::UserAgent;
use HTML::TreeBuilder::XPath;
use Term::ANSIColor;

# Prompt for target website URL
print colored("Enter your target website URL: ", "cyan");
my $target_url = <STDIN>;
chomp $target_url;

# Send request to the target website
my $ua = LWP::UserAgent->new;
my $response = $ua->get($target_url);
die "Failed to fetch $target_url: ", $response->status_line unless $response->is_success;

# Parse HTML content
my $content = $response->decoded_content;
my $tree = HTML::TreeBuilder::XPath->new;
$tree->parse($content);

# Check for vulnerability
my $vulnerable = 0;
my @scripts = $tree->findvalues('//script[@src]');
foreach my $script (@scripts) {
    if ($script =~ /^https?:\/\/(?!$target_url)/i) {
        $vulnerable = 1;
        last;
    }
}

# Display vulnerability status
print "\n";
if ($vulnerable) {
    print colored("The target website is vulnerable to HTML5 Cross-Origin Messaging!\n", "red");
} else {
    print colored("The target website is not vulnerable to HTML5 Cross-Origin Messaging.\n", "green");
}

# Display how to test the vulnerability
print colored("\nTo test the vulnerability, you can use the following steps:\n", "cyan");
print colored("1. Open Developer Tools in your browser (usually by pressing F12).\n", "yellow");
print colored("2. Navigate to the 'Console' tab.\n", "yellow");
print colored("3. Execute JavaScript code similar to:\n", "yellow");
print colored("   var iframe = document.createElement('iframe');\n", "yellow");
print colored("   iframe.src = 'https://attacker.com';\n", "yellow");
print colored("   document.body.appendChild(iframe);\n", "yellow");
print colored("4. If you see a request to 'https://attacker.com' being made from the target website, it is vulnerable.\n\n", "yellow");

# Cleanup
$tree->delete;

File: hacking_with_python/HTML5_Cross-Origin_Messaging.py
import requests
from bs4 import BeautifulSoup
from termcolor import colored

def test_html5_cors(target_url):
    print(colored("[*] Testing HTML5 Cross-Origin Messaging for: " + target_url, "blue"))

    # Send a GET request to the target URL
    try:
        response = requests.get(target_url)
        if response.status_code != 200:
            print(colored("[!] Error: Unable to access the target website.", "red"))
            return
    except requests.RequestException as e:
        print(colored("[!] Error: Unable to connect to the target website.", "red"))
        print(e)
        return

    # Parse the HTML content
    soup = BeautifulSoup(response.text, 'html.parser')
    iframe_tags = soup.find_all('iframe')

    # Check if any iframe tags with 'sandbox' attribute found
    vulnerable = False
    for tag in iframe_tags:
        if tag.get('sandbox'):
            print(colored("[!] Vulnerability Found: Cross-Origin Messaging may be enabled.", "red"))
            vulnerable = True
            break

    if not vulnerable:
        print(colored("[*] Target is not vulnerable to HTML5 Cross-Origin Messaging.", "green"))
        return

    print(colored("[*] How to Test the Vulnerability:", "blue"))
    print(colored("1. Create an HTML file with a message and host it on a different domain.", "yellow"))
    print(colored("2. Include this HTML file in an iframe on the target website.", "yellow"))
    print(colored("3. Use JavaScript to listen for messages from the iframe.", "yellow"))
    print(colored("4. If you receive messages from the iframe, the website is vulnerable.", "yellow"))

if __name__ == "__main__":
    target_website = input(colored("Enter your target website url: ", "blue"))
    test_html5_cors(target_website)
